File: kotlin/src/main/kotlin/benchmarks/Nbody.kt
package benchmarks

import Benchmark
import kotlin.math.sqrt

class Nbody : Benchmark() {
    private val bodies = initialBodies()
    private var v1 = 0.0

    companion object {
        private const val SOLAR_MASS = 4.0 * Math.PI * Math.PI
        private const val DAYS_PER_YEAR = 365.24
    }

    private class Planet(
        var x: Double,
        var y: Double,
        var z: Double,
        vx: Double,
        vy: Double,
        vz: Double,
        mass: Double,
    ) {
        var vx = vx * DAYS_PER_YEAR
        var vy = vy * DAYS_PER_YEAR
        var vz = vz * DAYS_PER_YEAR
        val mass = mass * SOLAR_MASS

        fun moveFromI(
            bodies: Array<Planet>,
            dt: Double,
            i: Int,
        ) {
            for (j in i until bodies.size) {
                val b2 = bodies[j]
                val dx = x - b2.x
                val dy = y - b2.y
                val dz = z - b2.z

                val distance = sqrt(dx * dx + dy * dy + dz * dz)
                val mag = dt / (distance * distance * distance)
                val bMassMag = mass * mag
                val b2MassMag = b2.mass * mag

                vx -= dx * b2MassMag
                vy -= dy * b2MassMag
                vz -= dz * b2MassMag
                b2.vx += dx * bMassMag
                b2.vy += dy * bMassMag
                b2.vz += dz * bMassMag
            }

            x += dt * vx
            y += dt * vy
            z += dt * vz
        }
    }

    private fun initialBodies() =
        arrayOf(
            Planet(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0),
            Planet(
                4.84143144246472090e+00,
                -1.16032004402742839e+00,
                -1.03622044471123109e-01,
                1.66007664274403694e-03,
                7.69901118419740425e-03,
                -6.90460016972063023e-05,
                9.54791938424326609e-04,
            ),
            Planet(
                8.34336671824457987e+00,
                4.12479856412430479e+00,
                -4.03523417114321381e-01,
                -2.76742510726862411e-03,
                4.99852801234917238e-03,
                2.30417297573763929e-05,
                2.85885980666130812e-04,
            ),
            Planet(
                1.28943695621391310e+01,
                -1.51111514016986312e+01,
                -2.23307578892655734e-01,
                2.96460137564761618e-03,
                2.37847173959480950e-03,
                -2.96589568540237556e-05,
                4.36624404335156298e-05,
            ),
            Planet(
                1.53796971148509165e+01,
                -2.59193146099879641e+01,
                1.79258772950371181e-01,
                2.68067772490389322e-03,
                1.62824170038242295e-03,
                -9.51592254519715870e-05,
                5.15138902046611451e-05,
            ),
        )

    private fun energy(bodies: Array<Planet>): Double {
        var e = 0.0
        val nbodies = bodies.size

        for (i in 0 until nbodies) {
            val b = bodies[i]
            e += 0.5 * b.mass * (b.vx * b.vx + b.vy * b.vy + b.vz * b.vz)
            for (j in i + 1 until nbodies) {
                val b2 = bodies[j]
                val dx = b.x - b2.x
                val dy = b.y - b2.y
                val dz = b.z - b2.z
                val distance = sqrt(dx * dx + dy * dy + dz * dz)
                e -= (b.mass * b2.mass) / distance
            }
        }
        return e
    }

    private fun offsetMomentum(bodies: Array<Planet>) {
        var px = 0.0
        var py = 0.0
        var pz = 0.0

        for (b in bodies) {
            px += b.vx * b.mass
            py += b.vy * b.mass
            pz += b.vz * b.mass
        }

        val b = bodies[0]
        b.vx = -px / SOLAR_MASS
        b.vy = -py / SOLAR_MASS
        b.vz = -pz / SOLAR_MASS
    }

    override fun prepare() {
        offsetMomentum(bodies)
        v1 = energy(bodies)
    }

    override fun run(iterationId: Int) {
        repeat(1000) {
            for (i in bodies.indices) {
                bodies[i].moveFromI(bodies, 0.01, i + 1)
            }
        }
    }

    override fun checksum(): UInt {
        val v2 = energy(bodies)
        return (Helper.checksumF64(v1) shl 5) and Helper.checksumF64(v2)
    }

    override fun name(): String = "CLBG::Nbody"
}

File: nim/src/benchmarks/nbody.nim
import std/[math]
import ../benchmark
import ../helper

type
  Planet = object
    x, y, z: float
    vx, vy, vz: float
    mass: float

  Nbody* = ref object of Benchmark
    resultVal: uint32
    bodies: seq[Planet]
    v1: float

const
  SOLAR_MASS = 4.0 * PI * PI
  DAYS_PER_YEAR = 365.24

proc newPlanet(x, y, z, vx, vy, vz, mass: float): Planet =
  Planet(
    x: x, y: y, z: z,
    vx: vx * DAYS_PER_YEAR,
    vy: vy * DAYS_PER_YEAR,
    vz: vz * DAYS_PER_YEAR,
    mass: mass * SOLAR_MASS
  )

proc moveFromI(bodies: var seq[Planet], dt: float, start: int) =
  let b = bodies[start]
  for i in start+1..<bodies.len:
    let b2 = bodies[i]
    let dx = b.x - b2.x
    let dy = b.y - b2.y
    let dz = b.z - b2.z

    let distance = sqrt(dx*dx + dy*dy + dz*dz)
    let mag = dt / (distance * distance * distance)
    let bMassMag = b.mass * mag
    let b2MassMag = b2.mass * mag

    bodies[start].vx -= dx * b2MassMag
    bodies[start].vy -= dy * b2MassMag
    bodies[start].vz -= dz * b2MassMag
    bodies[i].vx += dx * bMassMag
    bodies[i].vy += dy * bMassMag
    bodies[i].vz += dz * bMassMag

  bodies[start].x += dt * bodies[start].vx
  bodies[start].y += dt * bodies[start].vy
  bodies[start].z += dt * bodies[start].vz

proc newNbody(): Benchmark =
  Nbody(resultVal: 0, v1: 0.0)

method name(self: Nbody): string = "CLBG::Nbody"

method prepare(self: Nbody) =

  self.bodies = @[
    newPlanet(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0),
    newPlanet(4.84143144246472090e+00, -1.16032004402742839e+00, -1.03622044471123109e-01,
              1.66007664274403694e-03, 7.69901118419740425e-03,
              -6.90460016972063023e-05,
              9.54791938424326609e-04),
    newPlanet(8.34336671824457987e+00, 4.12479856412430479e+00, -4.03523417114321381e-01,
              -2.76742510726862411e-03, 4.99852801234917238e-03,
              2.30417297573763929e-05,
              2.85885980666130812e-04),
    newPlanet(1.28943695621391310e+01, -1.51111514016986312e+01, -2.23307578892655734e-01,
              2.96460137564761618e-03, 2.37847173959480950e-03,
              -2.96589568540237556e-05,
              4.36624404335156298e-05),
    newPlanet(1.53796971148509165e+01, -2.59193146099879641e+01, 1.79258772950371181e-01,
              2.68067772490389322e-03, 1.62824170038242295e-03,
              -9.51592254519715870e-05,
              5.15138902046611451e-05)
  ]

  var px, py, pz = 0.0
  for b in self.bodies:
    px += b.vx * b.mass
    py += b.vy * b.mass
    pz += b.vz * b.mass

  self.bodies[0].vx = -px / SOLAR_MASS
  self.bodies[0].vy = -py / SOLAR_MASS
  self.bodies[0].vz = -pz / SOLAR_MASS

  var e = 0.0
  let nbodies = self.bodies.len
  for i in 0..<nbodies:
    let b = self.bodies[i]
    e += 0.5 * b.mass * (b.vx*b.vx + b.vy*b.vy + b.vz*b.vz)
    for j in i+1..<nbodies:
      let b2 = self.bodies[j]
      let dx = b.x - b2.x
      let dy = b.y - b2.y
      let dz = b.z - b2.z
      let distance = sqrt(dx*dx + dy*dy + dz*dz)
      e -= (b.mass * b2.mass) / distance

  self.v1 = e

method run(self: Nbody, iteration_id: int) =
  for j in 0..<1000:
    for i in 0..<self.bodies.len:
      self.bodies.moveFromI(0.01, i)

method checksum(self: Nbody): uint32 =

  var e = 0.0
  let nbodies = self.bodies.len
  for i in 0..<nbodies:
    let b = self.bodies[i]
    e += 0.5 * b.mass * (b.vx*b.vx + b.vy*b.vy + b.vz*b.vz)
    for j in i+1..<nbodies:
      let b2 = self.bodies[j]
      let dx = b.x - b2.x
      let dy = b.y - b2.y
      let dz = b.z - b2.z
      let distance = sqrt(dx*dx + dy*dy + dz*dz)
      e -= (b.mass * b2.mass) / distance

  let v2 = e
  let checksum1 = checksumF64(self.v1)
  let checksum2 = checksumF64(v2)
  (checksum1 shl 5) and checksum2

registerBenchmark("CLBG::Nbody", newNbody)
